Supporting VM changes for extensionification

cwillisf opened the pull request:
### Dependency

This PR and LLK/scratch-gui#4790 should be merged at about the same time, as they make a small-but-important change to the way extension block information is communicated between the VM and the GUI.

### Proposed Changes

This is a grab-bag of supporting changes for extensionification, including:
* Support for interpreting extension block info through a Blockly mutation (late / "dynamic") instead of in the VM (immediate / "static")
  * This makes it possible to have two different blocks with the same opcode but different attributes in some other way, such as two variable blocks or custom procedure call blocks.
  * This also opens the door to removing some of the `scratch-blocks` / Blockly knowledge from `scratch-vm` if we decide to remove the "static" path.
* Loading a "core" extension will now place the extension blocks in the correct location in the toolbox instead of at the bottom. Non-"core" extensions will still load at the bottom.
* It's now possible to make core extension blocks look the same as the non-extension core blocks:
  * Extension blocks which don't have an extension icon will no longer be "inflated" using the `scratch_extension` Blockly extension.
  * Colors specified by an extension are now honored.
* Extension metadata will now be updated when changing the current editing target by selecting a different sprite/stage.

### Reason for Changes

It might make sense to review and/or merge this as one PR, or it might make sense to break it up into more than one. Either way, this PR can serve as a way to keep tabs on the current state of the `e16n` branch.

### Test Coverage

* Added `test/unit/engine_mutation-adapter.js` to test the new mutation adapter features.
* Modified `test/integration/internal-extension.js`, `test/unit/engine_blocks.js`, and `test/unit/extension_conversion.js` to cover new or modified block and category information.

Commit:
pass `blockInfo` to extension methods

Sometimes a single extension method needs to service several different
instances of the same opcode. This can happen with variables or custom
procedures, for example. This change allows the extension method to
inspect the `blockInfo` for instance data, including arbitrary
extension-specific properties if necessary.

Commit:
add test for extension methods receiving `blockInfo`

Commit:
add category info to extension add & update events

Commit:
retrieve blockInfo from args when isDynamic is set

Commit:
embed extension blockInfo into block XML

Commit:
adjust getBlocksXML to return categories separately

before: getBlocksXML returns one big XML string
after: getBlocksXML returns an array of {id,xml}, one entry per category

Commit:
add "scratch_extension" only if a block has an icon

kchadha committed:
Fix serialization of blockInfo mutation property to XML

Commit:
fix extension block color application

kchadha committed:
Add a runtime event to track when the toolbox extension blocks need updating.

kchadha committed:
Update comment for new extension update event

kchadha committed:
Merge pull request #2141 from kchadha/toolbox-update

Add a runtime event for requesting a toolbox update

towerofnix commented on docs/extensions.md:
```diff
@@ -152,6 +152,10 @@ class SomeBlocks {
 
+            // Core extensions only: override the default extension block colors.
+            color1: '#FF8C1A',
+            color2: '#DB6E00',
```
Should `color3` be here?

Also, when this is called `color1`, `color2`, `color3`, to me it sounds like it's just yelling to be in the format `colors: [first, second, third]`. But that takes an additional comment making it explicit that this is an *ordered* array, not just a set of colors. So, instead, I think the names `colorPrimary`, `colorSecondary`, and `colorTertiary` would be better, along with a description of what those mean (which will be easier to remember than color1/2/3!) and maybe a link to [the color specification](https://github.com/LLK/scratch-blocks/wiki/Colors).

Also, yeah, internally I think these [*are*](https://github.com/LLK/scratch-vm/pull/2161/commits/91f0d59be078f8c120b5429352f284ce87bc8c6b#diff-38d875e1d50b3158dc96404b65275a54R785) given the names color1/color2/color3. But (assuming that can't easily be changed immediately) it's still future proofing to give more specific names to the colors in extensions. Making a change to better names later would mean either all future extension code requiring updates, or including code in the VM to support both color1/2/3 and primary/secondary/tertiary formats (which makes any future code use *either* of those formats, which is sure to be confusing).

cwillisf replied:
Yeah - I've gone back and forth with how I feel about which names to use and whether to use an array. I feel like `colorPrimary`, `colorSecondary`, and `colorTertiary` don't offer much advantage since they basically mean the same as `color1`, `color2`, and `color3` and are harder to spell.

Still, I agree that these names aren't great either. Ideally I'd like more semantically-meaningful names. I'll think on this a bit more...

Also: I agree that a link to the color spec would be a good idea. Thanks for the suggestion!

kchadha replied:
Hmmm... semantically meaningful names... maybe we can talk to @carljbowman about this?

I *think* the meanings of the colors are as follows:
color1/colorPrimary: the base color of the block (usually the most prevalent color on the block)
color2/colorSecondary: The color of shadows/droppable menus on the block (e.g. the dark green on the boolean blocks)
color3/colorTertiary: The color of the outline of the blocks

If these definitions are actually correct, maybe we could have something like:

baseColor, shadowColor (wouldn't actually want to use the word shadow here necessarily), and outlineColor...?

![image](https://user-images.githubusercontent.com/1786240/57096539-88340f80-6ce3-11e9-84f0-ad210c0054c8.png)

cwillisf replied:
I talked to @carljbowman about this a while ago but forgot to put notes here -- sorry!

Anyway, he suggested that we stick with non-semantic names. I probably won't capture his reasoning accurately but the way I understood it is that there are a bunch of semantically-disconnected uses of each color that are hard to capture in just a few words.

Also, apparently [that color spec](https://github.com/LLK/scratch-blocks/wiki/Colors) is very out-of-date to the point that @carljbowman wasn't sure we should be linking to it since it might point people in the wrong direction. To me that suggests we should either take it down (and not link to it here) or update it as a separate project (and add a link here after that).

Commit:
add tests for new e16n support features

kchadha commented on src/engine/mutation-adapter.js:
```diff
@@ -13,6 +13,10 @@ const mutatorTagToObject = function (dom) {
     for (const prop in dom.attribs) {
         if (prop === 'xmlns') continue;
         obj[prop] = decodeHtml(dom.attribs[prop]);
+        if (prop === 'blockinfo') {
```
```suggestion
        // Note: the capitalization of block info in the following lines is important. 
        // The lowercase is read in from xml which normalizes case. The VM uses camel case everywhere else.
        if (prop === 'blockinfo') {
```

kchadha commented on src/engine/runtime.js:
```diff
@@ -801,9 +816,7 @@ class Runtime extends EventEmitter {
-        const allBlocks = fieldTypeDefinitionsForScratch.concat(categoryInfo.blocks).concat(categoryInfo.menus);
-
-        this.emit(Runtime.EXTENSION_ADDED, allBlocks);
+        this.emit(Runtime.EXTENSION_ADDED, categoryInfo);
```
🎉

cwillisf committed:
call out capitalization change since it's important but easy to miss

Co-Authored-By: Karishma Chadha <kchadha@media.mit.edu>

kchadha commented on src/engine/runtime.js:
```diff
@@ -812,18 +825,16 @@ class Runtime extends EventEmitter {
                 this._fillExtensionCategory(categoryInfo, extensionInfo);
-                extensionBlocks = extensionBlocks.concat(categoryInfo.blocks, categoryInfo.menus);
+
+                this.emit(Runtime.BLOCKSINFO_UPDATE, categoryInfo);
```
Documenting a conversation from offline...
It looks like this change would make this line run multiple times per call to `_refreshExtensionPrimitives` but in practice, this change is the same as before, this emit runs only once per call to `_refreshExtensionPrimitives` because there should only ever be one category in `this._blockInfo` which matches the given extension id.

We should refactor this in the future to prevent confusion.

cwillisf replied:
Refactored in fd776025e5fe2cfd2b2d5f7fa004ea94985b1447

kchadha commented on src/engine/runtime.js:
```diff
@@ -1272,12 +1286,12 @@ class Runtime extends EventEmitter {
+            return {
+                id: categoryInfo.id,
+                xml: `<category name="${name}" id="${categoryInfo.id}" ${statusButtonXML} ${colorXML} ${menuIconXML}>${
+                    paletteBlocks.map(block => block.xml).join('')}</category>`
```
Is the `.join('')` here necessary or just for ease of reading?

Commit:
refactor _refreshExtensionPrimitives for clarity

The previous form made it harder to see that at most one extension
category is handled per call.

kchadha reviewed (approved):
LGTM!
